File: src/lib/ics.ts
interface IcsInput {
  uid: string
  start: Date
  end: Date
  summary: string
  description: string
  location?: string
  organizerEmail: string
  attendeeEmail: string
  attendeeName: string
}

const escapeText = (s: string) => s.replace(/[\\,;]/g, m => `\\${m}`).replace(/\n/g, '\\n')

export function buildIcs({
  uid,
  start,
  end,
  summary,
  description,
  location,
  organizerEmail,
  attendeeEmail,
  attendeeName,
}: IcsInput): string {
  const fmt = (d: Date) => d.toISOString().replace(/[-:]/g, '').split('.')[0] + 'Z'

  return [
    'BEGIN:VCALENDAR',
    'VERSION:2.0',
    'PRODID:-//AuditMind//Booking//EN',
    'CALSCALE:GREGORIAN',
    'METHOD:REQUEST',
    'BEGIN:VEVENT',
    `UID:${uid}`,
    `DTSTAMP:${fmt(new Date())}`,
    `DTSTART:${fmt(start)}`,
    `DTEND:${fmt(end)}`,
    `SUMMARY:${escapeText(summary)}`,
    `DESCRIPTION:${escapeText(description)}`,
    location ? `LOCATION:${escapeText(location)}` : '',
    `ORGANIZER;CN=Audit Mind:mailto:${organizerEmail}`,
    `ATTENDEE;CN=${escapeText(attendeeName)};RSVP=TRUE:mailto:${attendeeEmail}`,
    'STATUS:CONFIRMED',
    'SEQUENCE:0',
    'END:VEVENT',
    'END:VCALENDAR',
  ]
    .filter(Boolean)
    .join('\r\n')
}

File: booking/utils/ics.py
# Port of src/lib/ics.ts — RFC 5545 VEVENT builder.
import re
from dataclasses import dataclass
from datetime import datetime, timezone


def _escape_text(s: str) -> str:
    s = re.sub(r'[\\,;]', lambda m: '\\' + m.group(0), s)
    return s.replace('\n', '\\n')


def _fmt(dt: datetime) -> str:
    return dt.astimezone(timezone.utc).strftime('%Y%m%dT%H%M%SZ')


@dataclass
class IcsInput:
    uid: str
    start: datetime
    end: datetime
    summary: str
    description: str
    organizer_email: str
    attendee_email: str
    attendee_name: str
    location: str | None = None


def build_ics(data: IcsInput) -> str:
    lines = [
        'BEGIN:VCALENDAR',
        'VERSION:2.0',
        'PRODID:-//AuditMind//Booking//EN',
        'CALSCALE:GREGORIAN',
        'METHOD:REQUEST',
        'BEGIN:VEVENT',
        f'UID:{data.uid}',
        f'DTSTAMP:{_fmt(datetime.now(timezone.utc))}',
        f'DTSTART:{_fmt(data.start)}',
        f'DTEND:{_fmt(data.end)}',
        f'SUMMARY:{_escape_text(data.summary)}',
        f'DESCRIPTION:{_escape_text(data.description)}',
        f'LOCATION:{_escape_text(data.location)}' if data.location else '',
        f'ORGANIZER;CN=Audit Mind:mailto:{data.organizer_email}',
        f'ATTENDEE;CN={_escape_text(data.attendee_name)};RSVP=TRUE:mailto:{data.attendee_email}',
        'STATUS:CONFIRMED',
        'SEQUENCE:0',
        'END:VEVENT',
        'END:VCALENDAR',
    ]
    return '\r\n'.join(line for line in lines if line)
